pkg/prometheus/operator.go

```diff
@@ -394,78 +419,83 @@ func ListOptions(name string) api.ListOptions {
 // create new pods.
 //
 // TODO(fabxc): remove this once the PetSet controller learns how to do rolling updates.
-func (c *Operator) syncVersion(p *spec.Prometheus) error {
-	fmt.Println("sync version")
-	defer fmt.Println("sync version complete")
-	opts := ListOptions(p.Name)
+func (c *Operator) syncVersion(key string, p *spec.Prometheus) error {
 	podClient := c.kclient.Core().Pods(p.Namespace)
 
-Outer:
-	for {
-		pods, err := podClient.List(opts)
+	pods, err := podClient.List(ListOptions(p.Name))
+	if err != nil {
+		return err
+	}
+
+	// If the PetSet is still busy scaling, don't interfere by killing pods.
+	// We enqueue ourselves again to until the PetSet is ready.
+	if len(pods.Items) != int(p.Spec.Replicas) {
+		return fmt.Errorf("scaling in progress")
+	}
+	if len(pods.Items) == 0 {
+		return nil
+	}
+
+	var oldPods []*v1.Pod
+	allReady := true
+	// Only proceed if all existing pods are running and ready.
+	for _, pod := range pods.Items {
+		ready, err := k8sutil.PodRunningAndReady(pod)
 		if err != nil {
-			return err
-		}
-		if len(pods.Items) == 0 {
-			return nil
-		}
-		for _, cp := range pods.Items {
-			ready, err := k8sutil.PodRunningAndReady(cp)
-			if err != nil {
-				return err
-			}
-			if !ready {
-				time.Sleep(200 * time.Millisecond)
-				continue Outer
-			}
+			c.logger.Log("msg", "cannot determine pod ready state", "err", err)
 		}
-		var pod *v1.Pod
-		for _, cp := range pods.Items {
-			if !strings.HasSuffix(cp.Spec.Containers[0].Image, p.Spec.Version) {
-				pod = &cp
-				break
+		if ready {
+			// TODO(fabxc): detect other fields of the pod template that are mutable.
+			if !strings.HasSuffix(pod.Spec.Containers[0].Image, p.Spec.Version) {
+				oldPods = append(oldPods, &pod)
 			}
+			continue
 		}
-		if pod == nil {
-			return nil
-		}
-		if err := podClient.Delete(pod.Name, nil); err != nil {
-			return err
-		}
+		allReady = false
 	}
-}
 
-func (c *Operator) deletePrometheus(p *spec.Prometheus) error {
-	// Update the replica count to 0 and wait for all pods to be deleted.
-	psetClient := c.kclient.Apps().PetSets(p.Namespace)
+	if len(oldPods) == 0 {
+		return nil
+	}
+	if !allReady {
+		return fmt.Errorf("waiting for pods to become ready")
```

Comment:
Yes, I figured the same initially. 

Core controllers just seem to re-enqueue on error too though. Just that their rate limiting is a bit more sophisticated with exponential backoff and eventually dropping an item fully.

Just doing it always is the safe route initially I think.